ILSpy.ReadyToRun/ReadyToRunLanguage.cs

```diff
@@ -0,0 +1,163 @@
+					// TODO: Indexing
+					foreach (RuntimeFunction runtimeFunction in m.RuntimeFunctions) {
+						WriteCommentLine(output, m.SignatureString);
+						byte[] code = new byte[runtimeFunction.Size];
+						for (int i = 0; i < runtimeFunction.Size; i++) {
+							code[i] = reader.Image[reader.GetOffset(runtimeFunction.StartAddress) + i];
+						}
+						DecoderFormatterExample(output, code, bitness, (ulong)runtimeFunction.StartAddress);
+						output.WriteLine();
+					}
+				}
+			}
+		}
+
+		private void DecoderFormatterExample(ITextOutput output, byte[] exampleCode, int exampleCodeBitness, ulong exampleCodeRIP)
+		{
+			// TODO: Decorate the disassembly with Unwind, GC and debug info
+			// You can also pass in a hex string, eg. "90 91 929394", or you can use your own CodeReader
+			// reading data from a file or memory etc
+			var codeBytes = exampleCode;
+			var codeReader = new ByteArrayCodeReader(codeBytes);
+			var decoder = Decoder.Create(exampleCodeBitness, codeReader);
+			decoder.IP = exampleCodeRIP;
+			ulong endRip = decoder.IP + (uint)codeBytes.Length;
+
+			// This list is faster than List<Instruction> since it uses refs to the Instructions
+			// instead of copying them (each Instruction is 32 bytes in size). It has a ref indexer,
+			// and a ref iterator. Add() uses 'in' (ref readonly).
+			var instructions = new InstructionList();
+			while (decoder.IP < endRip) {
+				// The method allocates an uninitialized element at the end of the list and
+				// returns a reference to it which is initialized by Decode().
+				decoder.Decode(out instructions.AllocUninitializedElement());
+			}
+
+			// Formatters: Masm*, Nasm*, Gas* (AT&T) and Intel* (XED)
+			// TODO: DecompilationOptions?
+			var formatter = new NasmFormatter();
+			formatter.Options.DigitSeparator = "`";
+			formatter.Options.FirstOperandCharIndex = 10;
+			var tempOutput = new StringBuilderFormatterOutput();
+			// Use InstructionList's ref iterator (C# 7.3) to prevent copying 32 bytes every iteration
+			foreach (var instr in instructions) {
+				// Don't use instr.ToString(), it allocates more, uses masm syntax and default options
+				formatter.Format(instr, tempOutput);
+				output.Write(instr.IP.ToString("X16"));
+				output.Write(" ");
+				int instrLen = instr.ByteLength;
+				int byteBaseIndex = (int)(instr.IP - exampleCodeRIP);
+				for (int i = 0; i < instrLen; i++)
+					output.Write(codeBytes[byteBaseIndex + i].ToString("X2"));
+				int missingBytes = 10 - instrLen;
+				for (int i = 0; i < missingBytes; i++)
+					output.Write("  ");
+				output.Write(" ");
+				output.WriteLine(tempOutput.ToStringAndReset());
+			}
+		}
+
+		private class R2RAssemblyResolver : ILCompiler.Reflection.ReadyToRun.IAssemblyResolver
+		{
+			public bool Naked => false;
+
+			public bool SignatureBinary => false;
+
+			public bool InlineSignatureBinary => false;
+
+			public string FindAssembly(string name, string filename)
+			{
+				throw new NotImplementedException();
```

Comment:
Can you clarify what the R2R.IAssemblyResolver is used for?

Reply:
I have done some investigation on this today and figured I have a rather serious bug in my implementation. 

Not implementing this function will actually lead to failure to open any R2R image other than `System.Private.CoreLib.dll`. 

I am working on implementing this method now.